batch.py

```diff
@@ -0,0 +1,389 @@
+)
+
+# region SUBSET_VARIANTS
+
+
+def filter_variants(
+    mt_path: str,  # "mt/v7.mt"
+    samples: list[str],
+    output_mt_path: str,  # "tob_wgs_rv/densified_rv_only.mt"
+):
+    """Subset hail matrix table
+
+    Input:
+    joint call hail matrix table
+    set of samples for which we have scRNA-seq data
+
+    Output:
+    subsetted hail matrix table, containing only variants that:
+    1) are not ref-only, 2) biallelic, 3) meet QC filters, 4) are rare (MAF<5%)
+    """
+    # read hail matrix table object (WGS data)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # subset to relevant samples (samples we have scRNA-seq data for)
+    mt = mt.filter_cols(hl.set(samples).contains(mt.s))
+
+    # densify (can this be done at the end?)
+    mt = hl.experimental.densify(mt)
+
+    # filter out low quality variants and consider biallelic variants only (no multi-allelic, no ref-only)
+    mt = mt.filter_rows(  # check these filters!
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)  # QC
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt = mt.checkpoint(output_mt_path, overwrite=True)
+    logging.info(
+        f"Number of rare (freq<5%) and QC'd biallelic variants: {mt.count()[0]}"
+    )
+
+
+# endregion SUBSET_VARIANTS
+
+
+# region GET_GENE_SPECIFIC_VARIANTS
+
+
+def get_promoter_variants(
+    mt_path: str,  # checkpoint from function above
+    ht_path: str,
+    gene_details: dict[str, str],  # ouput of make_gene_loc_dict
+    window_size: int,
+    plink_output_prefix: str,  # 'tob_wgs_rv/pseudobulk_rv_association/plink_files/'
+):
+    """Subset hail matrix table
+
+    Input:
+    mt_path: path to already subsetted hail matrix table
+    gene_file: path to a tsv file with information on
+    the chrom, start and end location of genes (rows)
+
+    Output:
+    For retained variants, that are: 1) in promoter regions and
+    2) within 50kb up or down-stream of the gene body (or in the gene body itself)
+    (on top of all filters done above)
+
+    returns:
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (pre-filtered)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    gene_name = gene_details["gene_name"]
+
+    # get relevant chromosome
+    chrom = gene_details["chrom"]
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    left_boundary = max(1, int(gene_details["gene-start"]) - window_size)
```

Comment:
I thought in your demo file the columns were `start` and `end` 👀

Reply:
yes you are right, i had already misunderstood how this worked and i had renames the columns (i thought specifying it was the gene's start, vs e.g. the variant was useful) - but with the current syntax this is wrong, you're right!